### PR TITLE
fix(storage): alternative endpoints and Host header

### PR DESCRIPTION
**Pull request by coryan**
Some alternatives for `storage.googleapis.com` require setting the
`Host:` header to `storage.googleapis.com`. This change (I think)
simplifies the host-header computation: we let `libcurl` do its thing in
most cases (it pulls it from the URL), and we override it only in the
cases where we know it is needed.

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/6864)

### CODE REVIEW

File: google/cloud/storage/internal/curl_client.cc [hunk 1]
```diff
@@ -104,8 +104,22 @@ bool XmlEnabled() {
 
 }  // namespace
 
+std::string HostHeader(Options const& options, char const* service) {
+  // This function returns an empty string if In most cases libcurl will fill
+  // out the correct `Host: ` header based on the URL.  The main exception is
+  // `https://restricted.googleapis.com` Hard-code the Host header. Applications
```
Comment (devbww) [inline google/cloud/storage/internal/curl_client.cc:110]: s/Hard-code/where we hard-code/ perhaps?

File: google/cloud/storage/internal/curl_client.h [hunk 2]
```diff
@@ -31,6 +31,20 @@ inline namespace STORAGE_CLIENT_NS {
 namespace internal {
 class CurlRequestBuilder;
 
+/**
+ * Computes the `Host: ` header given the options.
```
Comment (devbww) [inline google/cloud/storage/internal/curl_client.h:35]: s/options/options and service/ perhaps?

**Review by devbww — APPROVED** [review]

File: google/cloud/storage/internal/curl_client.cc [hunk 3]
```diff
@@ -104,8 +104,22 @@ bool XmlEnabled() {
 
 }  // namespace
 
+std::string HostHeader(Options const& options, char const* service) {
+  // This function returns an empty string if In most cases libcurl will fill
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/storage/internal/curl_client.cc]: I changed all this text, hopefully it is clearer now.

File: google/cloud/storage/internal/curl_client.cc (hunk 1 above)
Comment (coryan) [reply google/cloud/storage/internal/curl_client.cc]: See above

File: google/cloud/storage/internal/curl_client.h (hunk 2 above)
Comment (coryan) [reply google/cloud/storage/internal/curl_client.h]: Done

**Review by devbww — APPROVED** [review]

**Pull request merged**